File: frontend/src/types/challenge.ts
/**
 * src/types/challenge.ts
 *
 * TypeScript types for the Challenge Room ONLY.
 * The existing src/types.ts (ClassicRoom) is NOT modified.
 *
 * Import these wherever ChallengeRoom.tsx needs types:
 *   import { UserRank, ChallengeQuestion, ... } from '../types/challenge';
 *
 * Then in ChallengeRoom.tsx remove the imports from '../types' that
 * belong to challenge (UserRank, ChallengeLevel, ChallengeQuestion,
 * ChallengeSessionState, Rank) and re-import from here.
 */

// ── Rank system ──────────────────────────────────────────────────────────

/** Rank letters from weakest to strongest */
export type Rank = 'E' | 'D' | 'C' | 'B' | 'A';

/** Challenge levels 1-5, same as difficulty 1-5 */
export type ChallengeLevel = 1 | 2 | 3 | 4 | 5;

/** Topic type — kept identical to ClassicRoom to reuse backend */
export type TopicType = 'History' | 'Geography' | 'Mixed';


// ── API response shapes (match backend Pydantic models exactly) ───────────

/**
 * GET /api/challenge/user/{user_id}/rank
 * Backend: UserRankOut
 */
export interface UserRank {
  current_rank    : Rank;
  rank_points     : number;
  available_levels: ChallengeLevel[];
  total_sessions  : number;
  total_questions : number;

  /** Convenience alias used by the frontend level-card display */
  level_access    : ChallengeLevel[];   // same as available_levels
  total_points    : number;             // alias for rank_points
}

/**
 * POST /api/challenge/start-session
 * Backend: StartSessionOut
 */
export interface StartSessionResponse {
  session_id      : string;
  current_level   : ChallengeLevel;
  rank_points     : number;
  available_levels: ChallengeLevel[];
  current_rank    : Rank;
  topic           : TopicType;
}

/**
 * POST /api/challenge/generate-question
 * Backend: ChallengeQuestionOut
 */
export interface ChallengeQuestion {
  id           : string;
  text         : string;
  options      : string[];
  explanation  : string;
  level        : ChallengeLevel;
  points_value : number;    // points for a correct answer at this level
  is_free_text : boolean;
}

/**
 * Forced level change signal returned from submit-answer.
 * Backend: ForceLevelChange
 */
export interface ForceLevelChange {
  direction: 'up' | 'down';
  reason   : string;
}

/**
 * POST /api/challenge/submit-answer
 * Backend: SubmitChallengeAnswerOut
 */
export interface SubmitAnswerResponse {
  id                : string | null;
  is_correct        : boolean;
  correct_answer    : string;
  explanation       : string;
  points_change     : number;       // signed: +7 or -4
  new_rank_points   : number;       // session running total
  new_level         : ChallengeLevel;
  streak_correct    : number;
  streak_wrong      : number;
  force_level_change: ForceLevelChange | null;
}

/**
 * POST /api/challenge/session/{id}/end
 * Backend: EndSessionOut
 */
export interface EndSessionResponse {
  session_id         : string;
  total_questions    : number;
  correct_answers    : number;
  total_points_earned: number;
  new_rank           : Rank;
  new_rank_points    : number;
  rank_changed       : boolean;
}


// ── Frontend session state (held in React useState) ──────────────────────

/**
 * Full in-memory session state managed by ChallengeRoom.tsx.
 * Not sent to the backend — it mirrors the backend DB row.
 */
export interface ChallengeSessionState {
  session_id    : string;
  topic         : TopicType;
  current_level : ChallengeLevel;
  current_rank  : Rank;
  rank_points   : number;             // session-level points
  streak_correct: number;
  streak_wrong  : number;
  questions     : ChallengeQuestion[];
  currentIndex  : number;
  score         : number;             // correct answer count
  pointsEarned  : number;             // alias for rank_points (used by UI)
  force_level_change: ForceLevelChange | null;
}

File: backend/schemas/challenge.py
"""
Challenge Room request/response models.

These models mirror the Challenge frontend contract in
frontend/src/types/challenge.ts and stay separate from Classic Room schemas.
"""

from __future__ import annotations
from typing import Literal, List, Optional
from pydantic import BaseModel, Field


# Shared Challenge literals.

RankLiteral = Literal["E", "D", "C", "B", "A"]
TopicType   = Literal["History", "Geography", "Mixed"]


# GET /api/challenge/user/{user_id}/rank

class UserRankOut(BaseModel):
    """What the frontend reads to show rank badge + available levels."""
    current_rank    : RankLiteral
    rank_points     : int
    available_levels: List[int]          # e.g. [1, 2, 3] for rank D
    total_sessions  : int = 0
    total_questions : int = 0

    model_config = {"populate_by_name": True}


# POST /api/challenge/start-session

class StartSessionRequest(BaseModel):
    user_id       : str
    topic         : TopicType
    starting_level: int = Field(ge=1, le=5)


class StartSessionOut(BaseModel):
    session_id      : str
    current_level   : int
    rank_points     : int               # always 0 at start
    available_levels: List[int]
    current_rank    : RankLiteral
    topic           : TopicType


# GET /api/challenge/session/{session_id}

class ChallengeSessionOut(BaseModel):
    session_id      : str
    user_id         : str
    topic           : str
    starting_level  : int
    current_level   : int
    rank_points     : int
    streak_correct  : int
    streak_wrong    : int
    total_questions : int
    correct_answers : int
    is_completed    : bool


# PATCH /api/challenge/session/{session_id}/change-level

class ChangeLevelRequest(BaseModel):
    direction: Literal["up", "down"]
    reason   : str = ""


class ChangeLevelOut(BaseModel):
    session_id : str
    new_level  : int
    direction  : str
    reason     : str


# POST /api/challenge/generate-question

class GenerateChallengeQuestionRequest(BaseModel):
    session_id: str
    user_id   : str
    topic     : TopicType
    level     : int = Field(ge=1, le=5)


class ChallengeQuestionOut(BaseModel):
    """
    Extends the classic QuestionOut with challenge-specific fields.
    correctAnswer is NOT returned to the frontend to prevent cheating.
    The answer is verified server-side on submit.
    """
    id           : str
    text         : str
    options      : List[str]
    explanation  : str
    level        : int
    points_value : int          # points if answered correctly at this level
    is_free_text : bool = False  # always False for now (MCQ only)

    model_config = {"populate_by_name": True}


# POST /api/challenge/submit-answer

class SubmitChallengeAnswerRequest(BaseModel):
    session_id  : str
    question_id : str
    user_id     : str
    answer      : str
    time_taken  : Optional[float] = None


class ForceLevelChange(BaseModel):
    direction: Literal["up", "down"]
    reason   : str


class SubmitChallengeAnswerOut(BaseModel):
    id                : Optional[str] = None
    is_correct        : bool
    correct_answer    : str
    explanation       : str
    points_change     : int         # actual signed value applied (+ or -)
    new_rank_points   : int         # session running total
    new_level         : int
    streak_correct    : int
    streak_wrong      : int
    force_level_change: Optional[ForceLevelChange] = None


# POST /api/challenge/session/{session_id}/end

class EndSessionOut(BaseModel):
    session_id          : str
    total_questions     : int
    correct_answers     : int
    total_points_earned : int
    new_rank            : RankLiteral
    new_rank_points     : int       # global cumulative after this session
    rank_changed        : bool      # True if rank letter changed


# Resolve Pydantic models eagerly during import.
UserRankOut.model_rebuild()
StartSessionRequest.model_rebuild()
StartSessionOut.model_rebuild()
ChallengeSessionOut.model_rebuild()
ChangeLevelRequest.model_rebuild()
ChangeLevelOut.model_rebuild()
GenerateChallengeQuestionRequest.model_rebuild()
ChallengeQuestionOut.model_rebuild()
SubmitChallengeAnswerRequest.model_rebuild()
SubmitChallengeAnswerOut.model_rebuild()
EndSessionOut.model_rebuild()
